paddle/operators/cos_sim_op.h

```diff
@@ -28,74 +28,114 @@ template <typename Place, typename T>
-    auto x = EigenMatrix<T>::From(*input_x, new_dims);
-    auto y = EigenMatrix<T>::From(*input_y, new_dims);
-    auto z = EigenMatrix<T>::From(*input_z);
-    auto x_norm = EigenMatrix<T>::From(*input_x_norm);
-    auto y_norm = EigenMatrix<T>::From(*input_y_norm);
-    auto dz = EigenMatrix<T>::From(*input_grad_z);
+    // convert Tensor to Eigen Tensor
+    int rows_x = in_x->dims()[0];
+    int rows_y = in_y->dims()[0];
+    int cols = framework::product(in_x->dims()) / rows_x;
+    auto x = EigenMatrix<T>::From(*in_x, framework::make_ddim({rows_x, cols}));
+    auto y = EigenMatrix<T>::From(*in_y, framework::make_ddim({rows_y, cols}));
+    auto z = EigenMatrix<T>::From(*in_z);
+    auto x_norm = EigenMatrix<T>::From(*in_x_norm);
+    auto y_norm = EigenMatrix<T>::From(*in_y_norm);
+    auto dz = EigenMatrix<T>::From(*in_grad_z);
 
-    Eigen::DSizes<int, 2> bcast(1, new_dims[1]);
+    // compute gradident
+    Eigen::DSizes<int, 2> bcast(1, cols);
     auto z_bcast = z.broadcast(bcast);
     auto dz_bcast = dz.broadcast(bcast);
-    auto place = context.GetEigenDevice<Place>();
     auto x_snorm_bcast = x_norm.square().eval().broadcast(bcast);
-    auto y_snorm_bcast = y_norm.square().eval().broadcast(bcast);
-    auto norm_prod_bcast = (x_norm * y_norm).eval().broadcast(bcast);
-    if (output_grad_x) {
-      output_grad_x->mutable_data<T>(context.GetPlace());
-      auto dx = EigenMatrix<T>::From(*output_grad_x, new_dims);
-      dx.device(place) =
-          dz_bcast * (y / norm_prod_bcast - z_bcast * x / x_snorm_bcast);
-    }
-    if (output_grad_y) {
-      output_grad_y->mutable_data<T>(context.GetPlace());
-      auto dy = EigenMatrix<T>::From(*output_grad_y, new_dims);
-      dy.device(place) =
-          dz_bcast * (x / norm_prod_bcast - z_bcast * y / y_snorm_bcast);
+    auto place = context.GetEigenDevice<Place>();
+    if (rows_x == rows_y) {
+      auto y_snorm_bcast = y_norm.square().eval().broadcast(bcast);
+      auto norm_prod_bcast = (x_norm * y_norm).eval().broadcast(bcast);
+      // compute dx
+      if (out_grad_x) {
+        out_grad_x->mutable_data<T>(context.GetPlace());
+        auto dx = EigenMatrix<T>::From(*out_grad_x,
+                                       framework::make_ddim({rows_x, cols}));
+        auto grad = y / norm_prod_bcast - z_bcast * x / x_snorm_bcast;
+        dx.device(place) = dz_bcast * grad;
+      }
+      // compute dy
+      if (out_grad_y) {
+        out_grad_y->mutable_data<T>(context.GetPlace());
+        auto dy = EigenMatrix<T>::From(*out_grad_y,
+                                       framework::make_ddim({rows_y, cols}));
+        auto grad = x / norm_prod_bcast - z_bcast * y / y_snorm_bcast;
+        dy.device(place) = dz_bcast * grad;
+      }
+    } else {
+      Eigen::DSizes<int, 2> bcast_row(rows_x, 1);
+      auto y_bcast =  y.broadcast(bcast_row);
+      auto y_snorm_bcast =
+          y_norm.square().eval().broadcast(bcast_row).eval().broadcast(bcast);
```

Comment:
Please merge the two ```broadcast``` operations to one like ```broadcast({rows, cols})```

Reply:
Done.